python/dgl/node_flow.py

```diff
@@ -0,0 +1,845 @@
+                for i in range(self.num_blocks):
+                    eid = utils.toindex(self.block_parent_eid(i))
+                    self._parent._edge_frame.update_rows(eid, self._edge_frames[i], inplace=True)
+            elif edge_embed_names is not None:
+                assert isinstance(edge_embed_names, list) \
+                        and len(edge_embed_names) == self.num_blocks, \
+                        "The specified embedding names should be the same as the number of flows."
+                for i in range(self.num_blocks):
+                    eid = utils.toindex(self.block_parent_eid(i))
+                    _update_frame(self._parent._edge_frame, edge_embed_names[i], eid,
+                                  self._edge_frames[i])
+
+    def map_to_parent_nid(self, nid):
+        """This maps the child node Ids to the parent Ids.
+
+        Parameters
+        ----------
+        nid : tensor
+            The node ID array in the NodeFlow graph.
+
+        Returns
+        -------
+        Tensor
+            The parent node id array.
+        """
+        return self._node_mapping.tousertensor()[nid]
+
+    def map_to_parent_eid(self, eid):
+        """This maps the child edge Ids to the parent Ids.
+
+        Parameters
+        ----------
+        nid : tensor
+            The edge ID array in the NodeFlow graph.
+
+        Returns
+        -------
+        Tensor
+            The parent edge id array.
+        """
+        return self._edge_mapping.tousertensor()[eid]
+
+    def layer_in_degree(self, layer_id):
+        """Return the in-degree of the nodes in the specified layer.
+
+        Parameters
+        ----------
+        layer_id : int
+            The layer Id.
+
+        Returns
+        -------
+        Tensor
+            The degree of the nodes in the specified layer.
+        """
+        return self._graph.in_degrees(utils.toindex(self.layer_nid(layer_id))).tousertensor()
+
+    def layer_out_degree(self, layer_id):
+        """Return the out-degree of the nodes in the specified layer.
+
+        Parameters
+        ----------
+        layer_id : int
+            The layer Id.
+
+        Returns
+        -------
+        Tensor
+            The degree of the nodes in the specified layer.
+        """
+        return self._graph.out_degrees(utils.toindex(self.layer_nid(layer_id))).tousertensor()
+
+    def layer_nid(self, layer_id):
+        """Get the node Ids in the specified layer.
+
+        Parameters
+        ----------
+        layer_id : int
+            The layer to get the node Ids.
+
+        Returns
+        -------
+        Tensor
+            The node id array.
+        """
+        layer_id = self._get_layer_id(layer_id)
+        assert layer_id + 1 < len(self._layer_offsets)
+        start = self._layer_offsets[layer_id]
+        end = self._layer_offsets[layer_id + 1]
+        return F.arange(int(start), int(end))
+
+    def layer_parent_nid(self, layer_id):
```

Comment:
Do we need this API? To me it seems to be equal to:
```python
nflow = ...
nid = nflow.layer_nid(i)
parent_nid = nflow.map_to_parent_nid(nid)
```

Reply:
This is a more efficient way of implementing it. Getting the parent nid of a layer is a common operation.